src/renderers/dom/shared/ReactDOMComponent.js

```diff
@@ -632,12 +633,21 @@ ReactDOMComponent.Mixin = {
           this._currentElement.type
         );
       }
+      var isCustomComponentTag = isCustomComponent(this._tag, props);
+      if (__DEV__ && isCustomComponentTag && !didWarnShadyDOM && el.shadyRoot) {
+        warning(
+          false,
+          'A component is using shady DOM. Using shady DOM with React can ' +
```

Comment:
This would need corresponding tests for a class, an old-style (`React.createClass`) component to make sure the message is correct.